### PR TITLE
materialized: Move https to a separate port for pgwire.

### LINKED ISSUE #12698: materialized: move HTTPS server to port 443
### Overview

`materialized` currently hosts its HTTPS server _and_ its pgwire server on the Materialize port (6875). There have been at least two requests to change it:

  * @bobbyiliev would like to access `materialized` from a Cloudflare worker, which only permits outbound traffic on ports 80/443.
  * @elindsey networking background leads him to deeply dislike the complexity introduced by hosting two protocols on the same port.

Historically, we've multiplexed these services on the same port to simplify operations for on-premise deployments: there's only one port users need to open in their firewall. Now that we're going cloud native, that argument no longer applies. 

So: I propose that we move Materialize's public HTTPS server to port 443.

### Design

Replace the `--listen-addr` argument to `materialized` with three new arguments:

  1. `--sql-listen-addr`: the port on which `materialized` listens for incoming pgwire connections
  2. `--http-listen-addr`: the port on which `materialized`  listens for incoming HTTP connections for public endpoints (e.g., /api/sql and the dataflow visualizer)
  3. `--internal-sql-listen-addr`: the port on which `materialized` listens for incoming pgwire connections from the cloud system. 
  4. `--internal-http-listen-addr`: the port on which `materialized` listens for incoming HTTP connections to access internal APIs (e.g., Prometheus metrics and memory debugging). This will replace the existing `--metrics-listen-addr` option.

Ports 1 and 2 will be authenticated/encrypted if requested via `--frontegg-auth`/`--tls-mode`. Ports 3 and 4 will always be unauthenticated and are not intended to be publicly exposed in Materialize Cloud.

Note that this design also incorporates the unauthenticated SQL port originally described in https://github.com/MaterializeInc/cloud/issues/3082#issuecomment-1132127898. Not strictly part of this issue, but it's intertwined with this issue, so easiest to tackle all at once.

### PR DESCRIPTION
**Pull request by wangandi**
### Motivation

  * This PR adds a known-desirable feature.

Closes #12698.

### Tips for reviewer

The code in mux.rs is now just common code for handling pgwire and http connections because I deleted the mux struct. I have assumed that after separating the ports that I no longer need to check for a handshake match.

### Testing

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.

Do I need to add a test for `--internal-sql-listen-addr`? I don't see any tests for the now `--internal-http-listen-addr`. Are there pre-existing tests for authenticated/encrypted connections to `--sql-listen-addr` and `--http-listen-addr`?

I have manually confirmed that the ports work.

### Release notes

This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):

The HTTP port is now separate from the pgwire port. The HTTP port has default port number 6876.

### CODE REVIEW

File: src/materialized/src/mux.rs [hunk 1]
```diff
@@ -7,51 +7,43 @@
...
-
-/// A mux routes incoming TCP connections to a dynamic set of connection
-/// handlers. It enables serving multiple protocols over the same port.
-///
-/// Connections are routed by sniffing the first several bytes sent over the
-/// wire and matching them against each handler, in order. The first handler
-/// to match the connection will be invoked.
-pub struct Mux {
-    handlers: Handlers,
-}
```
Comment (guswynn) [inline src/materialized/src/mux.rs:33]: getting rid of this seems right, we probably want to rename the file though!

File: src/sqllogictest/src/runner.rs [hunk 2]
```diff
@@ -580,13 +580,15 @@ impl Runner {
                 tracing: TracingCliArgs::default(),
             },
             secrets_controller: None,
-            listen_addr: SocketAddr::new(IpAddr::V4(Ipv4Addr::LOCALHOST), 0),
+            sql_listen_addr: SocketAddr::new(IpAddr::V4(Ipv4Addr::LOCALHOST), 0),
+            http_listen_addr: SocketAddr::new(IpAddr::V4(Ipv4Addr::LOCALHOST), 0),
```
Comment (guswynn) [inline src/sqllogictest/src/runner.rs:584]: will this not fail to setup the second listener if the ports are the same?

File: src/testdrive/src/action.rs [hunk 3]
```diff
@@ -93,9 +93,12 @@ pub struct Config {
     pub backoff_factor: f64,
 
     // === Materialize options. ===
-    /// The connection parameters for the materialized instance that testdrive
+    /// The pgwire connection parameters for the materialized instance that testdrive
     /// will connect to.
     pub materialized_pgconfig: tokio_postgres::Config,
+    /// The port for the public endpoints of the materialized instance that
+    /// testdrive will connect to via HTTP.
+    pub materialized_http_port: u16,
```
Comment (guswynn) [inline src/testdrive/src/action.rs:101]: do you need the port and the http_addr below? or can the addr just include the port?

**Review by guswynn — COMMENTED** [review]
rust parts look reasonable to me!

File: misc/python/materialize/mzcompose/services.py [hunk 4]
```diff
@@ -73,17 +74,38 @@ def __init__(
...
+        config_ports = [sql_port, 5432, *extra_ports]
+        default_port_commands = [
+            f"--sql-listen-addr 0.0.0.0:{guest_sql_port}",
+            f"--http-listen-addr 0.0.0.0:{guest_http_port}",
+        ]
+
+        if isinstance(image, str) and ":v" in image:
+            version = image.split(":v")[1]
+            version_nums = version.split(".")
+            if int(version_nums[0]) == 0 and int(version_nums[1]) <= 26:
```
[the remark this answers is not in the document]
Comment (necaris) [reply misc/python/materialize/mzcompose/services.py:100]: `distutils.version` should already be installed and be "good enough" -- https://stackoverflow.com/a/11887885 has other options too?

File: src/materialized/ci-slim/Dockerfile [hunk 5]
```diff
@@ -20,4 +20,4 @@ COPY materialized /usr/local/bin/
 
 USER materialize
 
-ENTRYPOINT ["tini", "--", "materialized", "--listen-addr=0.0.0.0:6875"]
+ENTRYPOINT ["tini", "--", "materialized", "--sql-listen-addr=0.0.0.0:6875", "--http-listen-addr=0.0.0.0:6875"]
```
Comment (necaris) [inline src/materialized/ci-slim/Dockerfile:23]: Does `--http-listen-addr` need to be `6876` here?

File: src/sqllogictest/src/runner.rs (hunk 2 above)
Comment (wangandi) [reply src/sqllogictest/src/runner.rs:584]: I originally wondered this too, but it turns out that when you set an address to port 0, the os will dynamically assign an available port. https://www.lifewire.com/port-0-in-tcp-and-udp-818145

File: misc/python/materialize/mzcompose/services.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (wangandi) [reply misc/python/materialize/mzcompose/services.py:100]: Thanks! Indeed we already have `packaging.version` installed.

File: src/materialized/ci-slim/Dockerfile (hunk 5 above)
Comment (wangandi) [reply src/materialized/ci-slim/Dockerfile:23]: Good catch. I have no idea why this did not result in a test failure.

File: src/testdrive/src/action.rs (hunk 3 above)
Comment (wangandi) [reply src/testdrive/src/action.rs:101]: Are you asking about replacing the lines below (155-156)
```
    materialized_sql_addr: String,
    materialized_http_addr: String,
```
with
```
    materialized_host: String,
    materialized_sql_port: String,
    materialized_http_port: String
```
?

We could do the replacement if you think it looks neater. It's several more characters of refactoring across all testdrive tests.

File: src/materialized/src/lib.rs [hunk 6]
```diff
@@ -370,7 +377,7 @@ async fn serve_stash<S: mz_stash::Append + 'static>(
     .await?;
 
     // Listen on the third-party metrics port if we are configured for it.
-    if let Some(addr) = config.metrics_listen_addr {
+    if let Some(addr) = config.internal_http_listen_addr {
         let metrics_registry = config.metrics_registry.clone();
         task::spawn(|| "metrics_server", {
```
Comment (benesch) [inline src/materialized/src/lib.rs:382]:
```suggestion
        task::spawn(|| "internal_http_server", {
```

File: src/materialized/src/lib.rs [hunk 7]
```diff
@@ -370,7 +377,7 @@ async fn serve_stash<S: mz_stash::Append + 'static>(
...
             let server = http::MetricsServer::new(metrics_registry);
```
Comment (benesch) [inline src/materialized/src/lib.rs:383]: Let's rename this to `http::InternalServer`.

File: src/materialized/src/mux.rs (hunk 1 above)
Comment (benesch) [reply src/materialized/src/mux.rs:33]: Ah we probably want to get rid of the file entirely. I think it's fine to leave as is in this PR and then I'll swing by and get rid of it entirely in a fast follow. #12891 tracks that work.

File: src/materialized/src/lib.rs [hunk 8]
```diff
@@ -380,58 +387,77 @@ async fn serve_stash<S: mz_stash::Append + 'static>(
...
-            // TODO(benesch): replace with `listener.incoming()` if that is
-            // restored when the `Stream` trait stabilizes.
-            let mut incoming = TcpListenerStream::new(listener);
-            mux.serve(incoming.by_ref().take_until(drain_tripwire))
+            let mut incoming = TcpListenerStream::new(sql_listener);
+            pgwire_server
+                .serve(incoming.by_ref().take_until(sql_drain_tripwire))
+                .await;
+        }
+    });
```
Comment (benesch) [inline src/materialized/src/lib.rs:416]: We need another one of these for the internal pgwire server. It should look identical except it always sets `tls` and `frontegg` to `None`.

**Comment by wangandi** [comment]
I think this PR is ready to be taken out of draft mode!

Only open questions are in the testing section of the description.

**Comment by benesch** [comment]
> Do I need to add a test for --internal-sql-listen-addr? I don't see any tests for the now --internal-http-listen-addr.

I think let's do it in a follow up so we can get this merged!

> Are there pre-existing tests for authenticated/encrypted connections to --sql-listen-addr and --http-listen-addr?

Yes! In `src/materialized/tests`. See `test_auth` for example.

**Review by benesch — APPROVED** [review]
🙌🏽 thanks very much!

**Pull request merged**